### PR TITLE
Gutenboarding - Test Dynamic Previewing Designs on Hover

### LINKED ISSUE #39248: Gutenboarding: try rendering design picker grid dynamically
Background pbAok1-cS-p2

image

- How performant it is? If it's not performant enough, what can we do to make it more performant?
  - Can we render only on hover?
  - Can we render only thumbnails that are currently visible on the page?
  - Can we render thumbnails without big images?
- Do we have what we need coming from the API?
- What kind of header/footer we can render? 
- Can we allow scrolling the whole template in thumbnail vs just show what's cropped in the thumbnail?
- Can we inject some Global styles CSS into each thumbnail?

Let's merge the functionality only behind an additional feature flag for now.

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
For summary of these `BlockPreview` draft experiments, please visit `pbAok1-lN-p2`.

### Changes proposed in this Pull Request

This branch tests dynamic previewing in the design step of gutenboarding by using `BlockPreview` when the design is hovered. See #39450 for testing BlockPreview for all designs simultaneously, #39496 for 3 at a time, and #39526 for testing global font settings.

**TLDR:**   We should be able to run a dynamic preview in Gutenboarding if we limit to one preview at a time and the main blockers to investigate further are supporting 3rd party blocks and loading all required styles and settings.  Previews are best used as done in Starter Page Templates, where one previewer is used to then load selections.  Loading different previewers on hover may be limiting on lower end devices.

preview-on-hover

**More Info**
Currently the previews render broken layouts due to lack of 3rd party block support and missing theme styles and settings.  However, they are running and rendering some images so they are set up to give us a good idea about what could be possible in future iterations of Gutenboarding designs.

There is a small but noticable load time when the items are hovered or focused, but I think this idea of rendering a dynamic preview on hover is a possibility if needed.  However, this is not advisable for lower end devices and I think having scrollable static images per vertical / design will serve us better for a step like this.

### Testing instructions

* Test live: https://calypso.live/?branch=try/gutenboarding-hover-dynamic-preview
* Navigate to the Gutenboarding design step and Hover (or a11y focus) designs for selection.

Fixes #39248

### CODE REVIEW

**Comment by ockham** [comment]
> There is a small but noticable load time when the items are hovered or focused, but I think this idea of rendering a dynamic preview on hover is definitely a possibility if needed. However, I think having scrollable static images per vertical / design will serve us better for a step like this.

Would it be possible (and make sense) to provide a bit of UI feedback while loading? Like a very thin loading bar on top of the preview, or even changing the mouse cursor while hovering (is that still an accepted UX pattern? :sweat_smile: )

**Comment by ockham** [comment]
> Currently the previews render broken layouts due to jetpack and coblocks not being supported in Gutenboarding. However, they are running and rendering some images so they are set up to give us a good idea about what could be possible in future iterations of Gutenboarding designs.

Good point. This will require a more thorough discussion, but off the top of my head:

- Both JP blocks and CoBlocks are available (in minified JS/CSS) form on WP.com servers -- we could try loading them from there. (And we already consider them audited for the purposes of using them in Gutenberg.)
  - Even if this works, it means we can't tree-shake, so we might be loading extra kBs. This might be fine for a first iteration/proof-of-concept, but might be worth revisiting.
  - I think we `external`ise some of our biggest dependencies (`@wordpress/` stuff, React, lodash), so that should cut down on bundle size.
- If we need them in source form (to basically `import` their code; e.g. for tree-shaking)... Oh well. JP blocks are in the JP repo, CoBlocks in their own. We'd need to figure something clever out :sweat_smile:

**Comment by amamujee** [comment]
What are the performance implications of dynamic previews? Do we have a system where we can detect platform or device type and turn on/off more fancy features for lower-end devices or mobile devices?

**Comment by Addison-Stavlo** [comment]
> What are the performance implications of dynamic previews? 

Thats in part what we are trying to figure out with some of these drafts.  Really, they are not very performant and running more than one at a time is not advised (we have some other draft PRs and trying 3 at a time seems to bog down the system a noticeable amount on a good machine).

Il edit my TLDR at the top as it seems a bit misleading.  While we "can" do something like this on hover, loading previews in this manner is probably not a wise use of resources.  A selection step like this is better suited for static images per design + vertical selection that we can set up to scroll through in the same manner.  This would have much less system load and require less data.

A Block Preview like this seems best suited to when you would have only one previewer, with different block templates to load into it (and of course the NEED to preview highly customized content).  Here we are bundling the previewer with the design and spinning up different previewers for different designs, while this is better than running multiple previewers simultaneously the best use for the block previewer would be to have only one that previews different designs upon selection (such as the starter page templates picker in the wpcom block editor).

> Do we have a system where we can detect platform or device type and turn on/off more fancy features for lower-end devices or mobile devices?

Not that I am currently aware of.  In terms of this previewer I would consider it as a mini editor instance.  Any device that can run the editor should be able to run the previewer, as long as we respect it for what it is and use it in that manner.

**Comment by Addison-Stavlo** [comment]
> Would it be possible (and make sense) to provide a bit of UI feedback while loading? Like a very thin progress bar on top of the preview

UI Feedback while loading would be great.  But Im not sure how we could track the loading vs loaded state of the BlockPreview itself.  We could set up loading indicators around async tasks such as requesting designs, fetching resources, parsing blocks, etc.  However, setting some sort of loading indicator inside the block preview itself to display its specific loading progress might be a bit more tricky and I'm not really sure how we would go about that atm.

> or even changing the mouse cursor while hovering (is that still an accepted UX pattern? 😅 )

You mean like having the hand pointer turn into an hourglass icon while its loading?  That would be awesome and retro and I would be all for it.
